### PR TITLE
feature-gate #[must_use] for functions as `fn_must_use`

### PR DESCRIPTION
**Pull request by zackmdavis**
@eddyb I [was](https://github.com/rust-lang/rust/pull/43728#issuecomment-320854120) [dithering](https://github.com/rust-lang/rust/pull/43728#issuecomment-320856407) on this, but [your comment](https://github.com/rust-lang/rust/issues/43302#issuecomment-321174989) makes it sound like we do want a feature gate for this? Please advise.

r? @eddyb

### CODE REVIEW

File: src/test/compile-fail/feature-gate-fn_must_use.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// gate-test-fn_must_use
+
+#[must_use]
+fn use_it() -> bool {
```
Comment (kennytm) [inline src/test/compile-fail/feature-gate-fn_must_use.rs]: The test is written wrongly, it should add an `//~ ERROR #[must_use] on function is experimental` to one of these two lines (depends on which span is used to report the error).

File: src/test/compile-fail/feature-gate-fn_must_use.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// gate-test-fn_must_use
```
Comment (kennytm) [inline src/test/compile-fail/feature-gate-fn_must_use.rs]:
```
tidy error: /checkout/src/test/compile-fail/feature-gate-fn_must_use.rs:11: The file is already marked as gate test through its name, no need for a 'gate-test-fn_must_use' comment
```

**Comment by bluss** [comment]
It's a breaking change -- Rust 1.19 thinks `#[must_use]` on a function is ok to compile.

**Comment by zackmdavis** [comment]
> Rust 1.19 thinks #[must_use] on a function is ok to compile.

The behavior as of the current revision of this pull request (Travis is currently failing because of tidy and the feature-gate test, but we can fix that after we understand what we want) would be the same with respect to allowing `#[must_use]` on functions to compile, but the warning would only be emitted if the feature is enabled.

> it should add an `//~ ERROR #[must_use] on function is experimental`

@kennytm Are we allowed to make it a warning rather than a hard error? (To avoid it being a breaking change, see @bluss's comment.)

**Comment by kennytm** [comment]
@zackmdavis 

A [future-incompatible warning](https://github.com/rust-lang/rust/issues/34596) like E0122 can be issued. But that's up to the lang team to decide.

* If `fn_must_use` is disabled, `#[must_use]` on function is no-op (allowed or future-incompatible-linted), and ignoring the return value will produce no lint.
* If `fn_must_use` is enabled, `#[must_use]` on function does what we expected (allowed), and ignoring the return value will produce lint.

**Comment by alexcrichton** [comment]
Thanks for the PR @zackmdavis! I think @kennytm's suggestions are all that's needed to land this!

**Comment by zackmdavis** [comment]
(Updated—includes reworking some feature-gate functionality to be able to only emit a warning. On the other hand, if we decide we want a hard error (is there _really_ going to be much code in the wild that has a `#[must_use]` on a function even though it's currently a no-op?—does such silly code not deserve what it gets, or are our stability guarantees meant to be taken _that literally_?), a937dca could be excluded.)

File: src/test/compile-fail/feature-gate-fn_must_use.rs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+fn main() {
+    // FIXME: feature gates are tidy-required to have a specially named (or
+    // comment-annotated) compile-fail test (which MUST fail), but for
+    // backwards-compatibility reasons, we want `#[must_use]` on functions to
+    // be compilable even if the `fn_must_use` feature is absent.  The error
+    // below (which makes this compile-fail test fail) doesn't have anything to
+    // do with the feature-gating, but pragmatically if awkwardly solves this
+    // dilemma until a superior solution can be devised.
+    == //~ ERROR expected expression
```
Comment (alexcrichton) [inline src/test/compile-fail/feature-gate-fn_must_use.rs]: We also have `#[rustc_error]` used throughout the test suite which may help here?

**Review by alexcrichton — COMMENTED** [review]
Thanks! I wonder if it would be possible to gate as a "hard" error by default? That may help reduce the surrounding changes here!

**Comment by zackmdavis** [comment]
> I wonder if it would be possible to gate as a "hard" error by default?

Fine with me (pull request updated). A compatibility bullet point in the release notes seems appropriate, though.

**Comment by alexcrichton** [comment]
Oh oops sorry! I think I miscommunicated a bit. I think it's probably a good idea to avoid turning `#[must_use]` on functions into hard errors immediately, what I meant was that instead of specifying `Hard` to all the gating invocations that was just done by default, but `#[must_use]` still generated warnings.

I wonder if perhaps though there's a more targeted way of "deprecating" the `#[must_use]` attribute on functions?

**Comment by zackmdavis** [comment]
> what I meant was that instead of specifying `Hard` to all the gating invocations that was just done by default

Oh, right. I did do some defaulting for the macros (which can be variadic), but I was trying to avoid too much code duplication. Looking at it today, I'm sure a much better trade-off between avoiding duplication and avoiding polluting existing `feature_err` and `emit_feature_err` callsites with a new argument is possible. I'm out of town for a few days (RustConf!), but I can probably get to this on Monday?

**Comment by alexcrichton** [comment]
Yeah the behavior of the current patch is fine by me, thanks! I'd just recommend a littl erefactoring to avoid needing to pass `GateStrength::Hard` so often, and other than that r=me!

**Comment by zackmdavis** [comment]
@alexcrichton updated (the only Travis subjobs that haven't passed yet are all macOS stuff)

**Comment by kennytm** [comment]
@bors ping 

Hello, both [Travis](https://travis-ci.org/rust-lang/rust/builds/267860998) and [AppVeyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.4398) have finished successfully but nothing happens?

**Comment by kennytm** [comment]
@bors retry rollup 

* Spurious failure #43985
* Rollup because it is [known](https://github.com/rust-lang/rust/pull/43776#issuecomment-324587310) to pass CI.

**Comment by SimonSapin** [comment]
I think this should be reverted: https://github.com/rust-lang/rust/issues/44213#issuecomment-326277448

**Pull request merged**